gemini/sample-apps/accelerating_product_innovation/app/pages_utils/utils_insights.py

```diff
@@ -0,0 +1,203 @@
+"""
+This module provides functions for generating insights and searching
+relevant information from uploaded data.
+    This module:
+    * Retrieves relevant context from a vector database based on a user's
+      query.
+    * Leverages Gemini-Pro to generate a precise answer.
+    * Presents the answer along with top-matched context sources.
+"""
+
+# pylint: disable=E0401
+
+import json
+import os
+import re
+
+from app.pages_utils import utils_resources_store_embeddings
+from app.pages_utils.embedding_model import embedding_model_with_backoff
+from app.pages_utils.utils_get_llm_response import generate_gemini
+from dotenv import load_dotenv
+from google.cloud import storage
+import numpy as np
+import pandas as pd
+import streamlit as st
+
+load_dotenv()
+
+
+PROJECT_ID = os.getenv("PROJECT_ID")
+LOCATION = os.getenv("LOCATION")
+
+
+def parse_and_format_text(text: str) -> str:
+    """Utility funtion to parse and format the LLM generated response.
+
+    Args:
+        text (str): The text to parse and format.
+
+    Returns:
+        str: The parsed and formatted text.
+    """
+    bold_pattern = r"\*\*(.*?)\*\*"
+    formatted_text = re.sub(bold_pattern, r"<b>\1</b>", text)
+    return formatted_text
+
+
+def extract_bullet_points(text: str) -> list[str]:
+    """Extracts bullet points from the LLM generated text.
+
+    Args:
+        text (str): The text to extract bullet points from.
+
+    Returns:
+        list: A list of bullet points.
+    """
+    bullet_points = []
+    for line in text.splitlines():
+        if re.match(r"^[-*•\d]\s*", line):
+            bullet_points.append(parse_and_format_text(line))
+    return bullet_points
+
+
+def get_suggestions(state_key: str) -> None:
+    """Gets suggestions for the given state key.
+
+    Args:
+        state_key (str): The state key to get suggestions for.
+    """
+
+    if st.session_state.rag_search_term is None:
+        dff = st.session_state["processed_data_list"].head(2)
+        context = "\n".join(dff["content"].values)
+        prompt = f""" Context: \n {context} \n
+         generate 5 questions based on the given context
+      """
+    else:
+        context = st.session_state.rag_search_term
+        prompt = f""" Context: \n {context} \n
+            generate 5 questions based on the given context. The questions
+            should strictly be questions for further analysis of
+            {st.session_state.rag_search_term}
+        """
+    print("CONTEXT")
+    print(context)
+    gen_suggestions = generate_gemini(prompt)
+    st.session_state[state_key] = extract_bullet_points(gen_suggestions)
+
+
+def check_if_file_uploaded() -> pd.DataFrame:
```

Comment:
This should probably be named something different, like `get_embeddings_file_as_df` ?